File: algos/valid-parentheses/valid-parentheses.js
/**
 Given a string s containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.

An input string is valid if:

Open brackets must be closed by the same type of brackets.
Open brackets must be closed in the correct order.
Every close bracket has a corresponding open bracket of the same type.
 

Example 1:

Input: s = "()"
Output: true
Example 2:

Input: s = "()[]{}"
Output: true
Example 3:

Input: s = "(]"
Output: false
 */
var isValid = function (s) {
  // create stack
  // for each char in s
  // if char is an opener, push onto stack
  // otherwise
  // if top element of stack is corresponding opener tag,
  // pop stack
  // at end if stack is empty return true else false

  if (s.length === 1) return false;

  const stack = [];

  for (let i = 0; i < s.length; i++) {
    const char = s[i];
    if (char === '(') {
      stack.push(')');
    } else if (char === '{') {
      stack.push('}');
    } else if (char === "[") {
      stack.push(']');
    } else if (stack.length === 0 || stack.pop() !== char) {
      return false;
    }
  }

  return stack.length === 0;
};

console.log(isValid("()"))
console.log(isValid("()[]{}"))
console.log(isValid("(]"))
console.log(isValid("([)]")) // false

File: algos/valid-parentheses/valid-parentheses.go
var closingToOpening = map[byte]byte{
	')': '(',
	'}': '{',
	']': '[',
}

func isValid(str string) bool {
	// create stack
	// loop ever char in string
	// if char is opening brace push opposite onto stack
	// else if stack is empty or popped element is not current char return false

	// if stack is empty return true

	s := []byte(str)
	stack := make([]byte, 0)

	for _, ch := range s {
		matchingOpening, isClosing := closingToOpening[ch]

		if !isClosing {
			stack = append(stack, ch)
			continue
		}

		if len(stack) == 0 {
			return false
		}

		lastOpening := stack[len(stack)-1]
		stack = stack[:len(stack)-1]

		if lastOpening != matchingOpening {
			return false
		}
	}

	return len(stack) == 0
}